sql: optimize pg_catalog.pg_roles population to fix performance regression

rafiss opened the pull request:
Previously, populating `pg_catalog.pg_roles` made separate privilege lookups for each role option (CREATEROLE, CREATEDB, BYPASSRLS, and multiple REPLICATION variants). Each lookup resulted in queries to the role_options table, which is expensive in multi-region clusters.

This commit optimizes the privilege checks by first checking the in-memory roleOptionsJSON that was already fetched, and only falling back to the global privilege lookup if the role option is not present. This changes the check from O(n * privileges) database lookups to using data already available in memory for most cases.

Additionally, the role membership cache is updated to support tracking versions for the system.role_options and system.database_role_settings tables. This is done via a functional options pattern in RunAtCacheReadTS which allows callers to specify which additional table versions should be checked for cache invalidation. The information_schema implementation now passes both CheckRoleOptionsVersion and CheckDatabaseRoleSettingsVersion to ensure proper cache invalidation when those tables are modified. This cache invalidation is necessary because the privilege checks now rely on the roleOptions data fetched by the forEachRoleAtCacheReadTS query, and that data must reflect the current state of the underlying tables.

Fixes #152044

Release note (performance improvement): Fixed a performance regression in pg_catalog.pg_roles and pg_catalog.pg_authid by avoiding privilege lookups for each row in the table.

rafiss committed:
sql: optimize pg_catalog.pg_roles population to fix performance regression

Previously, populating `pg_catalog.pg_roles` made separate privilege
lookups for each role option (CREATEROLE, CREATEDB, BYPASSRLS, and
multiple REPLICATION variants). Each lookup resulted in queries to the
role_options table, which is expensive in multi-region clusters.

This commit optimizes the privilege checks by first checking the
in-memory roleOptionsJSON that was already fetched, and only falling
back to the global privilege lookup if the role option is not present.
This changes the check from O(n * privileges) database lookups to using
data already available in memory for most cases.

Additionally, the role membership cache is updated to support tracking
versions for the system.role_options and system.database_role_settings
tables. This is done via a functional options pattern in RunAtCacheReadTS
which allows callers to specify which additional table versions should
be checked for cache invalidation. The information_schema implementation
now passes both CheckRoleOptionsVersion and CheckDatabaseRoleSettingsVersion
to ensure proper cache invalidation when those tables are modified. This
cache invalidation is necessary because the privilege checks now rely on
the roleOptions data fetched by the forEachRoleAtCacheReadTS query, and
that data must reflect the current state of the underlying tables.

Fixes #152044

Release note (performance improvement): Fixed a performance regression
in pg_catalog.pg_roles and pg_catalog.pg_authid by avoiding privilege
lookups for each row in the table.

rafiss commented:
I used this script to test this:

<details>
<summary>Script</summary>

```
#!/usr/bin/env bash

# Copyright 2025 The Cockroach Authors.
#
# Use of this software is governed by the CockroachDB Software License
# included in the /LICENSE file.

# Script to reproduce issue #152044: increased latency when populating
# pg_catalog.pg_roles vtable in 25.3.
#
# Usage:
#   ./scripts/repro-152044.sh <cluster-name> [version|--binary <path>]
#
# Examples:
#   ./scripts/repro-152044.sh rafiss-repro-152044                            # Uses latest nightly
#   ./scripts/repro-152044.sh rafiss-repro-152044 v25.2.8                    # Uses specific release
#   ./scripts/repro-152044.sh rafiss-repro-152044 v25.3.1                    # Uses specific release
#   ./scripts/repro-152044.sh rafiss-repro-152044 --binary ./cockroach       # Uses local binary
#
# The script:
#   1. Creates a 9-node multi-region cluster (3 regions x 3 nodes)
#   2. Stages the specified binary (nightly or release)
#   3. Starts the cluster with multi-region configuration
#   4. Creates ~8500 roles to match the reported scenario
#   5. Measures the latency of SELECT * FROM pg_catalog.pg_roles

set -euo pipefail

if [[ $# -lt 1 ]]; then
    echo "Usage: $0 <cluster-name> [version|--binary <path>]"
    echo ""
    echo "Examples:"
    echo "  $0 rafiss-repro-152044                            # Uses latest nightly"
    echo "  $0 rafiss-repro-152044 v25.2.8                    # Uses specific release"
    echo "  $0 rafiss-repro-152044 v25.3.1                    # Uses specific release"
    echo "  $0 rafiss-repro-152044 --binary ./cockroach       # Uses local binary"
    exit 1
fi

CLUSTER="$1"
VERSION=""
LOCAL_BINARY=""

# Parse optional arguments
if [[ $# -ge 2 ]]; then
    if [[ "$2" == "--binary" ]]; then
        if [[ $# -lt 3 ]]; then
            echo "Error: --binary requires a path argument"
            exit 1
        fi
        LOCAL_BINARY="$3"
        if [[ ! -f "$LOCAL_BINARY" ]]; then
            echo "Error: binary not found: $LOCAL_BINARY"
            exit 1
        fi
    else
        VERSION="$2"
    fi
fi

# Cluster configuration: 3 regions x 3 nodes = 9 nodes
NODES=9
REGIONS="us-east1,us-west1,europe-west1"
ZONES="us-east1-b,us-east1-b,us-east1-b,us-west1-b,us-west1-b,us-west1-b,europe-west1-b,europe-west1-b,europe-west1-b"

# Number of roles to create (matching reported scenario of ~8579 rows)
NUM_ROLES=8500

echo "=== Reproducing issue #152044 ==="
echo "Cluster: $CLUSTER"
if [[ -n "$LOCAL_BINARY" ]]; then
    echo "Binary: $LOCAL_BINARY (local)"
elif [[ -n "$VERSION" ]]; then
    echo "Version: $VERSION (release)"
else
    echo "Version: latest nightly"
fi
echo ""

# Step 1: Create the cluster
echo "=== Step 1: Creating multi-region cluster ==="
roachprod create "$CLUSTER" \
    --nodes="$NODES" \
    --gce-zones="$ZONES" \
    --lifetime=8h

# Step 2: Stage the binary
echo ""
echo "=== Step 2: Staging binary ==="
if [[ -n "$LOCAL_BINARY" ]]; then
    echo "Uploading local binary $LOCAL_BINARY..."
    roachprod put "$CLUSTER" "$LOCAL_BINARY" cockroach
elif [[ -n "$VERSION" ]]; then
    echo "Staging release $VERSION..."
    roachprod stage "$CLUSTER" release "$VERSION"
else
    echo "Staging latest nightly..."
    roachprod stage "$CLUSTER" cockroach
fi

# Step 3: Start the cluster
echo ""
echo "=== Step 3: Starting cluster ==="
# Use locality flags for multi-region setup
# Node selector syntax: cluster:nodes
roachprod start "$CLUSTER":1-3 \
    --args="--locality=region=us-east1"

roachprod start "$CLUSTER":4-6 \
    --args="--locality=region=us-west1" \
    --skip-init

roachprod start "$CLUSTER":7-9 \
    --args="--locality=region=europe-west1" \
    --skip-init

echo "Waiting for cluster to stabilize..."
sleep 10

# Step 4: Initialize cluster and create roles
echo ""
echo "=== Step 4: Creating $NUM_ROLES roles ==="

# Create roles in batches of 1000 per transaction for better performance
BATCH_SIZE=1000
for ((i=0; i<NUM_ROLES; i+=BATCH_SIZE)); do
    end=$((i + BATCH_SIZE))
    if [[ $end -gt $NUM_ROLES ]]; then
        end=$NUM_ROLES
    fi

    # Build batch SQL with transaction wrapper
    # Disable autocommit_before_ddl to allow multiple DDL in one transaction
    sql="SET autocommit_before_ddl = off; BEGIN;"
    for ((j=i; j<end; j++)); do
        sql+="CREATE ROLE IF NOT EXISTS test_role_$j;"
    done
    sql+="COMMIT;"

    roachprod sql "$CLUSTER":1 -- -e "$sql" > /dev/null 2>&1

    echo "Created $end roles..."
done

echo "Created $NUM_ROLES roles."

# Step 5: Verify role count
echo ""
echo "=== Step 5: Verifying setup ==="
roachprod sql "$CLUSTER":1 -- -e "SELECT count(*) AS role_count FROM pg_catalog.pg_roles;"

# Step 6: Run the benchmark queries
echo ""
echo "=== Step 6: Measuring pg_catalog.pg_roles latency ==="

echo ""
echo "--- Query 1: SELECT * FROM pg_catalog.pg_roles (default permissions) ---"
roachprod sql "$CLUSTER":1 -- -e "
EXPLAIN ANALYZE SELECT * FROM pg_catalog.pg_roles;
" 2>&1 | tail -20

echo ""
echo "--- Query 2: Timing multiple runs ---"
for run in 1 2 3; do
    echo "Run $run:"
    start_time=$(date +%s.%N)
    roachprod sql "$CLUSTER":1 -- -e "SELECT count(*) FROM (SELECT * FROM pg_catalog.pg_roles);" 2>&1
    end_time=$(date +%s.%N)
    elapsed=$(echo "$end_time - $start_time" | bc)
    echo "  Elapsed: ${elapsed}s"
    echo ""
done

# Print cluster info for further investigation
echo ""
echo "=== Cluster Info ==="
echo "To connect to the cluster:"
echo "  roachprod sql $CLUSTER:1"
echo ""
echo "To view the admin UI:"
echo "  roachprod adminurl $CLUSTER:1"
echo ""
echo "To destroy the cluster when done:"
echo "  roachprod destroy $CLUSTER"
```

</details>

On the latest nightly build, latencies were around 20 seconds.

<details>
<summary>Logs from nightly</summary>

```
=== Step 6: Measuring pg_catalog.pg_roles latency ===

--- Query 1: SELECT * FROM pg_catalog.pg_roles (default permissions) ---
  DistSQL network usage: 0 B (0 messages)
  regions: us-east1
  kv cpu time: 0µs
  sql cpu time: 35ms
  isolation level: serializable
  priority: normal
  quality of service: regular

  • virtual table
    sql nodes: n1
    regions: us-east1
    actual row count: 8,503
    execution time: 23.1s
    sql cpu time: 35ms
    table: pg_roles@primary
(21 rows)

Time: 23.151s

Connection to 34.26.252.5 closed.

--- Query 2: Timing multiple runs ---
Run 1:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.26.252.5' (ED25519) to the list of known hosts.
  count
---------
   8504
(1 row)

Time: 23.424s

Connection to 34.26.252.5 closed.
  Elapsed: 25.612799000s

Run 2:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.26.252.5' (ED25519) to the list of known hosts.
  count
---------
   8504
(1 row)

Time: 20.664s

Connection to 34.26.252.5 closed.
  Elapsed: 24.798823000s

Run 3:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.26.252.5' (ED25519) to the list of known hosts.
  count
---------
   8504
(1 row)

Time: 20.418s

Connection to 34.26.252.5 closed.
  Elapsed: 22.798246000s
```

</details>

With this PR, they are around ~400ms:

<details>
<summary>Logs from this PR</summary>

```
=== Step 6: Measuring pg_catalog.pg_roles latency ===

--- Query 1: SELECT * FROM pg_catalog.pg_roles (default permissions) ---
  DistSQL network usage: 0 B (0 messages)
  regions: us-east1
  kv cpu time: 0µs
  sql cpu time: 10ms
  isolation level: serializable
  priority: normal
  quality of service: regular

  • virtual table
    sql nodes: n1
    regions: us-east1
    actual row count: 8,503
    execution time: 710ms
    sql cpu time: 10ms
    table: pg_roles@primary
(21 rows)

Time: 711ms

Connection to 34.138.132.75 closed.

--- Query 2: Timing multiple runs ---
Run 1:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.138.132.75' (ED25519) to the list of known hosts.
  count
---------
   8503
(1 row)

Time: 250ms

Connection to 34.138.132.75 closed.
  Elapsed: 2.557508000s

Run 2:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.138.132.75' (ED25519) to the list of known hosts.
  count
---------
   8503
(1 row)

Time: 246ms

Connection to 34.138.132.75 closed.
  Elapsed: 2.748614000s

Run 3:
WARN: cluster rafi-repro-152044 defaults to insecure, because it is in project cockroach-ephemeral
Warning: Permanently added '34.138.132.75' (ED25519) to the list of known hosts.
  count
---------
   8503
(1 row)

Time: 254ms

Connection to 34.138.132.75 closed.
  Elapsed: 2.591119000s
```

</details>

fqazi reviewed (approved):
@rafiss Nice optimization! One issue with the table ID

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> @fqazi reviewed 5 files and all commit messages, and made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160121#-:-OhGtkX02MZdVFqjC_Is:bbl6mh3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss).

___
*[`pkg/sql/rolemembershipcache/cache.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/160121#-OhGtgIT4d-AFNtOEKgL:-OhGtgIT4d-AFNtOEKgM:b66o39t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f067ef3aa8ac9319358b104d3a7af0879146264f/pkg/sql/rolemembershipcache/cache.go#L128)):*
> ```Go
> 	}
> 	if options.checkDatabaseRoleSettingsVersion {
> 		dbRoleSettingsDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)
> ```

Table ID here is incorrect


<!-- Sent from Reviewable.io -->

rafiss commented:
Trying to debug a failure in the backup/schemachange tests. I'm leaving notes to return to later.

It's hitting this error:
```
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_add_check_udf/post_commit_stage_1_of_3/restore_all_tables_in_database
    backup.go:575: error executing query="DROP DATABASE IF EXISTS \"defaultdb\" CASCADE" args=[]: pq: referenced descriptor ID 1: looking up ID 1: descriptor not found
```

It's linked to the SHOW BACKUP FROM LATEST query that's made by the test here: https://github.com/cockroachdb/cockroach/blob/0580abda806e0b196025e18047bfd3202862abdd/pkg/sql/schemachanger/sctest/backup.go#L341-L345

I added additional logging to show the stack when the error happens:
```
not found: goroutine 8923 [running]:
runtime/debug.Stack()
	GOROOT/src/runtime/debug/stack.go:26 +0x64
github.com/cockroachdb/cockroach/pkg/util/debugutil.Stack(...)
	pkg/util/debugutil/debugutil.go:63
github.com/cockroachdb/cockroach/pkg/sql/catalog.NewDescriptorNotFoundError(0x1)
	pkg/sql/catalog/errors.go:92 +0x20
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.requiredError({0x1082a2007?, 0x0?}, 0x1)
	pkg/sql/catalog/internal/catkv/catalog_query.go:219 +0x2e8
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.build({0x1082a2007, 0x3}, 0x1, 0x0, 0x1)
	pkg/sql/catalog/internal/catkv/catalog_query.go:189 +0x2b8
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogQuery.processDescriptorResultRow({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, 0x1, {0x1082a2007, 0x3}}, {{0x14005ed9c40?, 0x10ea99b40?, ...}, ...}, ...)
	pkg/sql/catalog/internal/catkv/catalog_query.go:116 +0x90
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogQuery.query({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, 0x1, {0x1082a2007, 0x3}}, {0x10a693da0, 0x14006dd9ae0}, ...)
	pkg/sql/catalog/internal/catkv/catalog_query.go:71 +0x264
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogReader.GetByIDs({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}}, {0x10a693da0, 0x14006dd9ae0}, 0x14003167810, {0x14005ed9c04, 0x1, ...}, ...)
	pkg/sql/catalog/internal/catkv/catalog_reader.go:419 +0x194
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.(*cachedCatalogReader).GetByIDs(0x1400594be60, {0x10a693da0, 0x14006dd9ae0}, 0x14003167810, {0x14005ed9c04, 0x1, 0x1}, 0x1, {0x1082a2007, 0x3}, ...)
	pkg/sql/catalog/internal/catkv/catalog_reader_cached.go:387 +0x250
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.mustGetDescriptorByIDs({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:453 +0xf4
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.mustGetDescriptorByID({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:439 +0x88
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.getForExpiration.func1({0x10a693da0, 0x14006dd9ae0}, 0x14003167810)
	pkg/sql/catalog/lease/storage.go:411 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0x14003167810, {0x10a693da0, 0x14006dd9ae0}, 0x14004cd3c78)
	pkg/kv/txn.go:1139 +0x74
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x10a693da0, 0x14006dd9ae0}, 0x14003167810, 0x1?)
	pkg/kv/db.go:1114 +0x3c
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x8?, {0x10a693da0, 0x14006dd9ae0}, 0x5?, 0x0?, 0x0, 0x14004cd3c78)
	pkg/kv/db.go:1077 +0xa8
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
	pkg/kv/db.go:1052
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.getForExpiration({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:405 +0x94
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).readOlderVersionForTimestamp(0x14001f33088, {0x10a693da0, 0x14006dd9ae0}, 0x1, {0x14006170600?, 0x1e03cde0?})
	pkg/sql/catalog/lease/lease.go:970 +0x19c
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).Acquire(0x14001f33088, {0x10a693da0, 0x14006dd9ae0}, {0x10a66b460, 0x14006170600}, 0x1)
	pkg/sql/catalog/lease/lease.go:2201 +0x118
github.com/cockroachdb/cockroach/pkg/sql/regions.NewCachedDatabaseRegionsAt({0x10a693da0, 0x14006dd9ae0}, 0x1400442a600?, 0x14001f33088, {0x14006dd9ae0?, 0x3a34840?})
	pkg/sql/regions/cached_db_regions.go:35 +0xc8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).checkDescriptorTwoVersionInvariant(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0})
	pkg/sql/conn_executor_exec.go:2275 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransactionInternal(0x14004ee4f08, {0x10a693da0?, 0x14006dd9ae0?})
	pkg/sql/conn_executor_exec.go:2457 +0x3a0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransaction(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0}, {0x10a6c0940, 0x14007bca3a0}, 0x14004cd44e8)
	pkg/sql/conn_executor_exec.go:2339 +0x19c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).handleAutoCommit(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0}, {0x10a6c0940, 0x14007bca3a0})
	pkg/sql/conn_executor_exec.go:4129 +0xec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenStateWithPausablePortal.func12({0x10a693da0?, 0x14006dd9ae0?})
	pkg/sql/conn_executor_exec.go:1618 +0xd4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenStateWithPausablePortal(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {{0x10a6c0940, 0x14007bca3a0}, {0x0, 0x0, 0x0}, {0x14002f16f2c, 0x2b}, ...}, ...)
	pkg/sql/conn_executor_exec.go:2149 +0x3818
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x10a693da0?, 0x14006dd9a90?})
	pkg/sql/conn_executor_exec.go:171 +0x6c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {0x10a6c0940, 0x14007bca3a0}, 0x0, 0x14004cd5670)
	pkg/sql/conn_executor_exec.go:4470 +0x310
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {{0x10a6c0940, 0x14007bca3a0}, {0x0, 0x0, 0x0}, {0x14002f16f2c, 0x2b}, ...}, ...)
	pkg/sql/conn_executor_exec.go:170 +0x508
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(...)
	pkg/sql/conn_executor.go:2426
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x14004ee4f08)
	pkg/sql/conn_executor.go:2431 +0x1948
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x14004ee4f08, {0x10a6947d0, 0x14006798f00}, 0x30000007d?, 0x140078ffc78?, 0x14001fa5c00?)
	pkg/sql/conn_executor.go:2339 +0x158
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0x14004d403c0?, {0x10a6947d0?, 0x14006798f00?}, {0x14001330a20?}, 0x2?, 0x0?)
	pkg/sql/conn_executor.go:1118 +0xa0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0x1400434e008, {0x10a6947d0, 0x14006798f00}, {0x0, 0x4, {0x1, {0x1082a1f92, 0x3}, {0x1400739d960, 0xf}, ...}, ...}, ...)
	pkg/sql/pgwire/conn.go:253 +0x304
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4()
	pkg/sql/pgwire/server.go:1260 +0xa4
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 8921
	pkg/sql/pgwire/server.go:1257 +0x424
```

I'm not sure if the `readOlderVersionForTimestamp` is expected above. This issue goes away when I remove the two new calls I added to lookup the system tables (`txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)` and `txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.DatabaseRoleSettingsTableID)`)

rafiss commented:
i fixed the above issue by changing the logic so it no longer invalidates the entire cache when role_options has changed; instead it just chooses a higher read timestamp.

bors r+
